File: python/tutorials/tour-of-restate-python/tour/app/cart_object.py
# Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
#
# This file is part of the Restate examples,
# which is released under the MIT license.
#
# You can find a copy of the license in the file LICENSE
# in the root directory of this repository or package or at
# https://github.com/restatedev/examples/

from restate.context import ObjectContext
from restate.object import VirtualObject

# <start_user_session>
cart = VirtualObject("CartObject")


@cart.handler("addTicket")
async def add_ticket(ctx: ObjectContext, ticket_id: str) -> bool:
    return True


@cart.handler()
async def checkout(ctx: ObjectContext) -> bool:
    return True


@cart.handler("expireTicket")
async def expire_ticket(ctx: ObjectContext, ticket_id: str):
    return
# <end_user_session>

File: java/tutorials/tour-of-restate-java/src/main/java/dev/restate/tour/part2/CartObject.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Java,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

package dev.restate.tour.part2;

import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.tour.auxiliary.CheckoutRequest;

import java.time.Duration;
import java.util.Set;

@VirtualObject
public class CartObject {
    // <start_add_ticket>
    @Handler
    public boolean addTicket(ObjectContext ctx, String ticketId) {

        boolean reservationSuccess = TicketObjectClient.fromContext(ctx, ticketId).reserve().await();

        if (reservationSuccess) {
            // !mark(1:3)
            CartObjectClient.fromContext(ctx, ctx.key())
                    .send()
                    .expireTicket(ticketId, Duration.ofMinutes(15));
        }

        return reservationSuccess;
    }
    // <end_add_ticket>

    @Handler
    public void expireTicket(ObjectContext ctx, String ticketId) {
        TicketObjectClient.fromContext(ctx, ticketId).send().unreserve();
    }

    @Handler
    public boolean checkout(ObjectContext ctx) {
        boolean checkoutSuccess = CheckoutServiceClient.fromContext(ctx)
                .handle(new CheckoutRequest("Mary", Set.of("seat2B")))
                .await();

        return checkoutSuccess;
    }

}
